### PR TITLE
Scaffolding of a FLASK app for WhatsApp analytics

### PR DESCRIPTION
**Pull request by Sachinbisht27**
#3

### Types of changes

- [ ] Bug fix (a change which fixes an issue)
- [x] New feature (a change which adds functionality)

### Short description of what this resolves:
Added changes
- [x] Application architecture
- [x] Database connectivity
- [x] Setup FLASK application
- [x] Updated README.MD for the installation guidelines
- [x] Logic to handle or process the webhook payload
- [x] Webhook Transaction log table to store the webhook's data
- [x] Google-cloud-logging for the monitoring of the service
- [x] CI checks for the better coding standards
- [x] Save the webhook transaction logs in the database as of now
- [x] Migrations for maintaining the database schemas through automation
- [ ] Create all the database models based on the database schema

### Checklist:

- [ ] I have performed a self-review of my own code.
- [x] The code follows the style guidelines of this project.
- [x] The code changes are passing the CI checks
- [ ] I have documented my code wherever required.
- [ ] The changes requires a change to the documentation.
- [ ] I have updated the documentation based on the my changes.
- [ ] I have added tests to cover my changes (if not applicable, please state why)
- [ ] All new and existing tests are passing.\

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+repos:
+-   repo: https://github.com/pre-commit/pre-commit-hooks
+    rev: v3.2.0
+    hooks:
+    -   id: trailing-whitespace
+    -   id: end-of-file-fixer
+    -   id: check-json
+    -   id: check-yaml
+    -   id: check-merge-conflict
+    -   id: check-added-large-files
```
Comment (Satendra-SR) [inline .pre-commit-config.yaml:10]: There are a few more hooks which can be helpful for our usecase
1. **flake8** - Purpose: Runs the Flake8 code linter to check for PEP 8 violations and other style issues.
2. **black** - Purpose: Applies the Black code formatter to format Python code according to PEP 8 standards.
3. **isort** - Purpose: Sorts import statements in Python files according to PEP 8 guidelines.
4. **mypy** - Purpose: Runs the MyPy static type checker to find type errors in Python code.
5. **docformatter** - Purpose: Formats docstrings in Python files to adhere to PEP 257 guidelines.

Comment (Sachinbisht27) [reply .pre-commit-config.yaml:10]: Added a few more hooks in the `pre-commit config`.

**Review by Satendra-SR — APPROVED** [review]

**Review by Satendra-SR — CHANGES_REQUESTED** [review]
Added some changes in the last review, got approved mistakenly

File: .github/workflows/pre-commit.yml [hunk 2]
```diff
@@ -0,0 +1,16 @@
+name: pre-commit
+
+on:
+  pull_request:
+  push:
+    branches: [main]
```
[the remark this answers is not in the document]
Comment (Sachinbisht27) [reply .github/workflows/pre-commit.yml]: Updated

File: README.md [hunk 3]
```diff
@@ -1,3 +1,41 @@
...
+    git clone https://github.com/DostEducation/whatsapp-webhook-analytics.git
+    ```
+2. Switch to project folder and setup the vertual environment
+    ```sh
+    cd whatsapp-webhook-analytics
+    python -m venv venv
+    ```
+3. Activate the virtual environment
+    ```sh
+    source ./venv/bin/activate
```
[the remark this answers is not in the document]
Comment (Sachinbisht27) [reply README.md:25]: Added command for win-os

File: README.md [hunk 4]
```diff
@@ -1,3 +1,41 @@
...
+    ```
+4. Install the dependencies:
+    ```sh
+    pip install -r requirements-dev.txt
```
[the remark this answers is not in the document]
Comment (Sachinbisht27) [reply README.md]: Merged

File: config.py [hunk 5]
```diff
@@ -0,0 +1,20 @@
+import os
+
+from dotenv import load_dotenv
+
+load_dotenv()
```
[the remark this answers is not in the document]
Comment (Sachinbisht27) [reply config.py]: Sure, @Satendra-SR Moved the `load_dotenv()` to the environment condition.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (Sachinbisht27) [reply .pre-commit-config.yaml:10]: Yes, @Satendra-SR. For the code formatting and increasing readability of the code.

File: migrations/env.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+            .replace("%", "%%")
+        )
+    except AttributeError:
+        return str(get_engine().url).replace("%", "%%")
+
+
+# add your model's MetaData object here
+# for 'autogenerate' support
+
+target_metadata = models.Base.metadata
```
Comment (Satendra-SR) [inline migrations/env.py:42]: Can we try this?

target_metadata = db.metadata

**Review by Satendra-SR — APPROVED** [review]
A few questions, else looks good to go

File: api/utils/loggingutils.py [hunk 7]
```diff
@@ -0,0 +1,18 @@
+import logging
+import os
+
+from google.cloud import logging as gcloud_logging
+
+from api import app
+
+logger = logging.getLogger()
+logging.basicConfig(level=logging.DEBUG)
```
[the remark this answers is not in the document]
Comment (Sachinbisht27) [reply api/utils/loggingutils.py]: Sure, updating.

File: config.py [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+FLASK_APP = os.environ.get("FLASK_APP", "development")
+
+if FLASK_APP == "development":
+    from os import path
+
+    from dotenv import load_dotenv
+
+    basedir = path.abspath(path.dirname(__file__))
+    load_dotenv(path.join(basedir, ".env"))
```
[the remark this answers is not in the document]
Comment (Sachinbisht27) [reply config.py]: Updated

**Pull request merged**